ugs-platform/ugs-platform-plugin-joystick/src/main/java/com/willwinder/ugs/nbp/joystick/service/JoystickServiceImpl.java

```diff
@@ -36,29 +44,88 @@ This file is part of Universal Gcode Sender (UGS).
+import static com.willwinder.ugs.nbp.joystick.Utils.ACTION_JOG_Y_PLUS;
+import static com.willwinder.ugs.nbp.joystick.Utils.ACTION_JOG_Z;
+import static com.willwinder.ugs.nbp.joystick.Utils.ACTION_MULTIPLY_FEED;
+import static com.willwinder.ugs.nbp.joystick.Utils.ACTION_START;
+import static com.willwinder.ugs.nbp.joystick.Utils.ACTION_STOP;
+import static com.willwinder.ugs.nbp.joystick.Utils.ACTION_Z_DOWN;
+import static com.willwinder.ugs.nbp.joystick.Utils.ACTION_Z_UP;
+
 public class JoystickServiceImpl implements JoystickService {
+
+    private static final Logger LOGGER = Logger.getLogger(JoystickServiceImpl.class.getSimpleName());
+
+    /**
+     * Treat value as zero if between 0.03 and -0.03.
+     */
+    public static final float AXIS_THRESHOLD = 0.03f;
+
     /**
      * Milliseconds to wait between reading joystick/gamepad values
      */
-    private static final int READ_DELAY_MILLISECONDS = 10;
-    private static final Logger LOGGER = Logger.getLogger(JoystickServiceImpl.class.getSimpleName());
+    private static final int READ_DELAY_MILLISECONDS = 1;
+
+    /**
+     * A version number for the settings so that we can handle version changes
+     */
+    private static final int SETTINGS_VERSION = 1;
 
     private final ControllerManager controllerManager;
     private final JoystickState joystickState;
     private final ExecutorService joystickReadThread;
+    private final ActionDispatcher joystickActionDispatcher;
+    private final ActionManager actionManager;
 
     private ControllerIndex currentController;
     private Set<JoystickServiceListener> listeners;
     private boolean isRunning;
+    private boolean isActionDispatcherActive = true;
 
     public JoystickServiceImpl() {
         joystickReadThread = Executors.newSingleThreadExecutor();
         controllerManager = new ControllerManager();
         joystickState = new JoystickState();
         listeners = new HashSet<>();
 
-        JoystickJogService joystickJogService = new JoystickJogService();
-        addListener(joystickJogService);
+        JogService jogService = CentralLookup.getDefault().lookup(JogService.class);
+        BackendAPI backendAPI = CentralLookup.getDefault().lookup(BackendAPI.class);
+        ContinuousJogWorker continuousJogWorker = new ContinuousJogWorker(backendAPI, jogService);
+
+        actionManager = new ActionManager();
+        actionManager.registerAction("continuousJogXAction", "Actions/Machine", new AnalogJogAction(continuousJogWorker, Axis.X));
+        actionManager.registerAction("continuousJogYAction", "Actions/Machine", new AnalogJogAction(continuousJogWorker, Axis.Y));
+        actionManager.registerAction("continuousJogZAction", "Actions/Machine", new AnalogJogAction(continuousJogWorker, Axis.Z));
+
+        joystickActionDispatcher = new ActionDispatcher(actionManager, continuousJogWorker);
+        addListener(joystickActionDispatcher);
+
+        if (Settings.getVersion() == 0) {
```

Comment:
Could you add a comment here that says something like... "Check if the settings have not been initialized".

Probably not the most common edge case, but this does get a little tricky if multiple controllers are used or the controller is upgraded. It's probably fine to ignore this, since the user would be the only one who knows that the device has changed.